### PR TITLE
Add typescript definition file

### LINKED ISSUE #54: TypeScript support
## Description

Provide TypeScript support

## Rationale

It would add type suggestions to the project configured with Typescript

### PR DESCRIPTION
**Pull request by enheit**
## Description

Add support for TypeScript typings
Closes #54

## Checklist

- [ ] I've added new tests for any features added
- [ ] I've added regression tests for any bugs fixed

### CODE REVIEW

**Comment by msrose** [comment]
Thanks for this! Going to take a closer look in the next day or so

**Comment by msrose** [comment]
The declaration file LGTM 👍 But I'm not very familiar with typescript - reading these docs https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html, we needs to include a "types" key in the package.json that points towards the definition file. Additionally, we publish the transpiled javascript to a `lib/` folder with babel - is it going to be a problem if babel sees .ts files in the src folder when it copies things into the lib folder? Do we need to alter the `build` script at all to make sure it publishes correctly?

**Comment by enheit** [comment]
> we need to include a "types" key in the package.json that points towards the definition file.   

Yes

> Do we need to alter the build script at all to make sure it publishes correctly?  

Yes

I added a new changes to the pull request (see 3eeb220). It adds one more additional step to the build process: copies file from `./src` folder to the `./lib` folder.

At least this approach [followed](https://github.com/reduxjs/reselect/blob/ac77610bbb0a3cab9b280ea5ea379c2387017446/package.json#L48) by the [reselect](https://github.com/reduxjs/reselect) library. They also store definition file in the project, next to the entry point.

File: package.json [hunk 1]
```diff
@@ -19,7 +20,7 @@
     "test": "jest",
     "lint": "eslint . --max-warnings 0",
     "fix": "eslint --fix .",
-    "build": "babel src -d lib",
+    "build": "babel src -d lib && ncp ./src/index.d.ts ./lib/index.d.ts",
```
Comment (msrose) [inline package.json]: Can we use the `--copy-files` option for babel here? https://babeljs.io/docs/en/babel-cli#copy-files That way we won't need the ncp dependency

**Comment by msrose** [comment]
Awesome, thanks for including the examples! I left a small comment on the PR which should fix the failing build, and after that is addressed I will do a quick manual test and we will hopefully be good to merge

**Comment by enheit** [comment]
@msrose I have found an issue with `import * as React from 'react'` on local environment. Thinking about how to fix it...

**Comment by msrose** [comment]
Did local testing to make sure the types file is copied over, LGTM 👍 ✅

**Comment by msrose** [comment]
@all-contributors please add @enheit for infrastructure

**Pull request merged**